### PR TITLE
refactor(ivy): LView is a proper linked list

### PR DESCRIPTION
**Pull request by benlesh**
- TView no longer stores childIndex
- LView now as CHILD_HEAD and CHILD_TAIL

TView used to store the head of the list, therefor all LViews had to have the same head, which is incorrect.

### CODE REVIEW

**Review by kara — CHANGES_REQUESTED** [review]
Looks good, but needs a test for the new behavior we want to support (e.g. using injector.get to get `ViewContainerRef`.

Also needs symbol test updates and linting.

**Review by gkalpak — COMMENTED** [review]
Isn't it possible to also remove views? Shouldn't `CHILD_HEAD`/`CHILD_TAIL` be updatd in that case?

File: packages/core/src/render3/node_manipulation.ts [hunk 1]
```diff
@@ -234,18 +235,19 @@ export function addRemoveViewFromContainer(
  */
 export function destroyViewTree(rootView: LView): void {
   // If the view has no children, we can clean it up and return early.
-  if (rootView[TVIEW].childIndex === -1) {
+  const rootChildHead = rootView[CHILD_HEAD];
+  if (!rootChildHead) {
     return cleanUpView(rootView);
   }
-  let viewOrContainer: LView|LContainer|null = getLViewChild(rootView);
+  let viewOrContainer: LView|LContainer|null = rootView[CHILD_HEAD];
```
[the remark this answers is not in the document]
Comment (benlesh) [reply packages/core/src/render3/node_manipulation.ts]: Hey, @gkalpak ... I guess I'm not sure what you mean (and the code has shifted a little since this comment), can you elaborate?

[the remark this answers is not in the document]
Comment (gkalpak) [reply packages/core/src/render3/node_manipulation.ts]: I mean you already have `rootChildHead = rootView[CHILD_HEAD]` a couple of lines above, so maybe you could reuse the same var. E.g.:

```diff
-let viewOrContainer: LView|LContainer|null = rootView[CHILD_HEAD];
+let viewOrContainer: LView|LContainer|null = rootChildHead;
```

...or even...

```diff
-const rootChildHead = rootView[CHILD_HEAD];
-if (!rootChildHead) {
+let viewOrContainer = rootView[CHILD_HEAD];
+if (!viewOrContainer) {
   return cleanUpView(rootView);
 }
-let viewOrContainer: LView|LContainer|null = rootView[CHILD_HEAD];
```

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -2056,8 +2054,10 @@ function generateInitialInputs(
  * @returns LContainer
  */
 export function createLContainer(
-    hostNative: RElement | RComment, currentView: LView, native: RComment,
+    hostNative: RElement | RComment | StylingContext | LView, currentView: LView, native: RComment,
     isForViewContainerRef?: boolean): LContainer {
+  ngDevMode && assertDomNode(native);
+  ngDevMode && assertLView(currentView, true);
```
Comment (AndrewKushnir) [inline packages/core/src/render3/instructions.ts]: I think it might be a bit hard to follow what `true` stands here for.
As an option, can we have 2 asserts here, like:
```
ngDevMode &&
    assertDefined(currentView, 'Current view must be defined') &&
    assertLView(currentView);
```
and keep just `assertLView` where required is false (default).
@benlesh what do you think?

**Comment by benlesh** [comment]
presubmit: https://test.corp.google.com/ui#id=OCL:232328029:BASE:232328145:1549306899223:9eaec735

File: packages/core/src/render3/instructions.ts (hunk 2 above)
Comment (benlesh) [reply packages/core/src/render3/instructions.ts]: I sort of see what you mean. But that also makes the assertions a lot "busier" when reading the code. I _think_ I'd prefer to keep the assertions a little more terse, myself, even though I'm generally in favor of more explicit, readable code. @mhevery what do you think?

Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: I am fine either way.

Comment (benlesh) [reply packages/core/src/render3/instructions.ts]: Can we do it in another PR? The FW-869 work I'm doing is built on top of this.

Comment (AndrewKushnir) [reply packages/core/src/render3/instructions.ts]: Sure, no problem, thanks the reply.

File: packages/core/src/render3/instructions.ts [hunk 3]
```diff
@@ -2500,7 +2502,7 @@ export function projection(nodeIndex: number, selectorIndex: number = 0, attrs?:
   const componentView = findComponentView(lView);
   const componentNode = componentView[HOST_NODE] as TElementNode;
   let nodeToProject = (componentNode.projection as(TNode | null)[])[selectorIndex];
-  let projectedView = componentView[PARENT] !;
+  let projectedView = getLViewParent(componentView) !;
```
Comment (kara) [inline packages/core/src/render3/instructions.ts]: No need to wrap this one - by definition, the component view is the view just below the component boundary, so it can't have a container on top.

File: packages/core/src/render3/instructions.ts [hunk 4]
```diff
@@ -2551,20 +2553,20 @@ export function projection(nodeIndex: number, selectorIndex: number = 0, attrs?:
...
-    lView: LView, adjustedHostIndex: number, state: T): T {
-  const tView = lView[TVIEW];
-  if (lView[TAIL]) {
-    lView[TAIL] ![NEXT] = state;
-  } else if (tView.firstTemplatePass) {
-    tView.childIndex = adjustedHostIndex;
+export function addToViewTree<T extends LView|LContainer>(lView: LView, lViewOrLContainer: T): T {
+  // TODO(benlesh/misko): This implementation is incorrect, because it always adds the LContainer to
+  // the end of the queue, which means if the developer asks for the LContainers out of order, the
+  // change detection will run out of order.
```
Comment (kara) [inline packages/core/src/render3/instructions.ts]: Maybe add clarification to comment about what "asks for" means?  Do you mean `injector.get(ViewContainerRef)`?  (Also is there a Jira ticket for this?)

File: packages/core/src/render3/interfaces/view.ts [hunk 5]
```diff
@@ -85,7 +85,7 @@ export interface LView extends Array<any> {
    * This is the "insertion" view for embedded views. This allows us to properly
    * destroy embedded views.
    */
-  [PARENT]: LView|null;
+  [PARENT]: LView|LContainer|null;
```
Comment (kara) [inline packages/core/src/render3/interfaces/view.ts]: Can you update the docs comment above to include containers?

File: packages/core/src/render3/node_manipulation.ts [hunk 6]
```diff
@@ -7,28 +7,32 @@
...
 
 export function getLContainer(tNode: TViewNode, embeddedView: LView): LContainer|null {
+  ngDevMode && assertLView(embeddedView, true);
   if (tNode.index === -1) {
     // This is a dynamically created view inside a dynamic container.
-    // If the host index is -1, the view has not yet been inserted, so it has no parent.
-    const containerHostIndex = embeddedView[CONTAINER_INDEX];
-    return containerHostIndex > -1 ? embeddedView[PARENT] ![containerHostIndex] : null;
+    const container = embeddedView[PARENT];
+    return isLContainer(container) ? container : null;
```
Comment (kara) [inline packages/core/src/render3/node_manipulation.ts]: The parent of a dynamic view isn't a container when it hasn't been attached yet, right?Think this needs a comment since it's not immediately obvious.

File: packages/core/src/render3/node_manipulation.ts [hunk 7]
```diff
@@ -7,28 +7,32 @@
...
   } else {
     // This is a inline view node (e.g. embeddedViewStart)
-    return embeddedView[PARENT] ![tNode.parent !.index] as LContainer;
+    const parent = embeddedView[PARENT] !;
+    return isLContainer(parent) ? parent : parent[tNode.parent !.index] as LContainer;
```
Comment (kara) [inline packages/core/src/render3/node_manipulation.ts]: When is the parent of an inline view not the container? It seems to me like this should be:

```ts
return embeddedView[PARENT];
```

File: packages/core/src/render3/node_manipulation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (kara) [reply packages/core/src/render3/node_manipulation.ts]: +1 to George's comment

File: packages/core/src/render3/util.ts [hunk 8]
```diff
@@ -9,15 +9,33 @@
...
+  ngDevMode && assertLView(lView, true);
+  const parent = lView[PARENT];
+  return isLContainer(parent) ? parent[PARENT] ! : parent;
+}
+
+/**
+ * Returns true if the value is an {@link LView}
+ * @param value the value to check
+ */
+export function isLView(value: any): value is LView {
```
Comment (kara) [inline packages/core/src/render3/util.ts]: I'm surprised we didn't have this yet ! Good add 👍

File: packages/core/src/render3/util.ts [hunk 9]
```diff
@@ -146,7 +166,7 @@ export function getRootView(target: LView | {}): LView {
   ngDevMode && assertDefined(target, 'component');
   let lView = Array.isArray(target) ? (target as LView) : readPatchedLView(target) !;
   while (lView && !(lView[FLAGS] & LViewFlags.IsRoot)) {
-    lView = lView[PARENT] !;
+    lView = lView[PARENT] !as LView;
```
Comment (kara) [inline packages/core/src/render3/util.ts]: We may walk up components that are inside embedded views

```suggestion
    lView = getLViewParent(lView) !;
```

Also, this looks almost identical to `getRootView` in `discovery_utils` (which you updated above as well). We should remove one of these as part of the cleanup.

**Review by kara — CHANGES_REQUESTED** [review]
Nice work, like how much more intuitive the data structures are.

File: packages/core/src/render3/util.ts (hunk 9 above)
Comment (benlesh) [reply packages/core/src/render3/util.ts]: Removed the other implementation of `getRootView`, and also a duplicate of `getRootContext`. :man_shrugging:

File: packages/core/src/render3/discovery_utils.ts [hunk 10]
```diff
@@ -96,28 +96,18 @@ export function getContext<T = {}>(element: Element): T|null {
...
-  let lView: LView = context.lView;
-  while (lView[PARENT] && lView[HOST] === null) {
+  let lView = context.lView;
+  let parent: LView|null;
+  ngDevMode && assertLView(lView);
+  while (lView[HOST] === null && (parent = getLViewParent(lView) !)) {
     // As long as lView[HOST] is null we know we are part of sub-template such as `*ngIf`
-    lView = lView[PARENT] !;
+    if (parent) {
+      lView = parent;
```
Comment (kara) [inline packages/core/src/render3/discovery_utils.ts]: If `parent` is falsy, does this mean that `lView` never gets re-assigned? Wouldn't this create an infinite loop? 🤔

File: packages/core/src/render3/instructions.ts (hunk 4 above)
Comment (kara) [reply packages/core/src/render3/instructions.ts]: ^ Still needs resolution

File: packages/core/src/render3/interfaces/view.ts (hunk 5 above)
Comment (kara) [reply packages/core/src/render3/interfaces/view.ts]: ^ Still needs resolution

File: packages/core/src/render3/interfaces/view.ts [hunk 11]
```diff
@@ -161,13 +162,25 @@ export interface LView extends Array<any> {
...
   [SANITIZER]: Sanitizer|null;
 
+  /**
+   * Reference to the first LView or LContainer beneath this LView in
+   * the hierarchy.
+   *
+   * Necessary to store this so views can traverse through their nested views
+   * to remove listeners and call onDestroy callbacks.
+   *
+   * For embedded views, we store the index of an LContainer's host rather than the first
```
[the remark this answers is not in the document]
Comment (kara) [reply packages/core/src/render3/interfaces/view.ts]: ^ Still needs resolution

**Review by kara — CHANGES_REQUESTED** [review]

**Comment by benlesh** [comment]
@kara The TODO in `addViewToTree` is actually addressed in another PR that's in-progress (for FW-869). So that will not be resolved in this PR.

**Comment by kara** [comment]
@benlesh To clarify, [the feedback that still needed resolution](https://github.com/angular/angular/pull/28382#discussion_r254950685) was that we need a better comment. I don't see why improving the comment needs to wait until the next PR.

> Maybe add clarification to comment about what "asks for" means? Do you mean `injector.get(ViewContainerRef)`? (Also is there a Jira ticket for this?)

File: packages/core/src/render3/discovery_utils.ts (hunk 10 above)
Comment (benlesh) [reply packages/core/src/render3/discovery_utils.ts]: If the `parent` is falsy, it wouldn't pass the `while` loop's test condition.

Comment (kara) [reply packages/core/src/render3/discovery_utils.ts]: So why do we need to check for it in line 104 then?

Comment (benlesh) [reply packages/core/src/render3/discovery_utils.ts]: The if statement was redundant. Removed.

File: packages/core/src/render3/interfaces/view.ts [hunk 12]
```diff
@@ -79,14 +79,18 @@ export interface LView extends Array<any> {
...
-   * The parent view is needed when we exit the view and must restore the previous
-   * `LView`. Without this, the render method would have to keep a stack of
+   * This may store an {@link LView} or {@link LContainer}.
+   *
+   * `LView` - The parent view. This is needed when we exit the view and must restore the previous
+   * LView. Without this, the render method would have to keep a stack of
    * views as it is recursively rendering templates.
    *
    * This is the "insertion" view for embedded views. This allows us to properly
    * destroy embedded views.
```
Comment (kara) [inline packages/core/src/render3/interfaces/view.ts]: Looks like this line should be removed since embedded views will not have the insertion view as a parent anymore.

**Review by kara — APPROVED** [review]
LGTM, aside from the outdated lines

**Comment by benlesh** [comment]
Caretaker: CI has been flakey for this one. Yarn refusing to update etc.

**Comment by IgorMinar** [comment]
google3 presumit fails... please update and remove the "cleanup" label when ready

**Comment by benlesh** [comment]
[new presubmit](https://test.corp.google.com/ui#id=OCL:232328029:BASE:235071797:1550788745754:83c6b2d9)

File: packages/core/src/debug/debug_node.ts [hunk 13]
```diff
@@ -7,13 +7,14 @@
...
-import {getComponent, getContext, getInjectionTokens, getInjector, getListeners, getLocalRefs, getViewComponent, isBrowserEvents, loadLContext, loadLContextFromNode} from '../render3/discovery_utils';
+import {getViewComponent} from '../render3/global_utils_api';
 import {TNode} from '../render3/interfaces/node';
 import {StylingIndex} from '../render3/interfaces/styling';
 import {LView, TData, TVIEW} from '../render3/interfaces/view';
 import {getProp, getValue, isClassBasedValue} from '../render3/styling/class_and_style_bindings';
 import {getStylingContext} from '../render3/styling/util';
-import {INTERPOLATION_DELIMITER, isPropMetadataString, renderStringify} from '../render3/util';
+import {getComponent, getContext, getInjectionTokens, getInjector, getListeners, getLocalRefs, isBrowserEvents, loadLContext, loadLContextFromNode} from '../render3/util/discovery_utils';
+import {INTERPOLATION_DELIMITER, isPropMetadataString, renderStringify} from '../render3/util/misc_utils';
```
Comment (AndrewKushnir) [inline packages/core/src/debug/debug_node.ts]: nit: I'm wondering if we could drop the `_utils` postfix (since it's already present in a path)?
`../render3/util/misc_utils` -> `../render3/util/misc`

Comment (benlesh) [reply packages/core/src/debug/debug_node.ts]: when the file is open in your IDE, it makes it easier to find... in particular `interfaces/view.ts` vs `view.ts` vs `utils/view.ts`... in many IDEs that would just be `view.ts` `view.ts` `views.ts`.

Comment (AndrewKushnir) [reply packages/core/src/debug/debug_node.ts]: Ok, thanks.

Comment (kara) [reply packages/core/src/debug/debug_node.ts]: ^ Yeah, I like that the `utils` is part of the file name.  `misc.ts` is pretty general on its own.

**Comment by IgorMinar** [comment]
this pr still fails on g3 presubmit. adding "cleanup" label.

root cause:
```
failed to load main  Error: Cyclical dependency between files:
/build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/view_utils.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/public_api.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/core.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/metadata.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/metadata/directives.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/jit/directive.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/misc_utils.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/view_traversal_utils.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/assert.ts -> /build/work/3b02cb54a0edf0b2b24d9ca8ccb8de16/google3/third_party/javascript/angular2/rc/packages/core/src/render3/util/view_utils.ts
```

**Review by mhevery — APPROVED** [review]

**Comment by benlesh** [comment]
@IgorMinar that's an old presubmit, it's passing now: 

https://test.corp.google.com/ui#id=OCL:232328029:BASE:235220976:1550861966438:b7aecde3
